### ISSUE TITLE
Multiple responses using oneOf attribute do not appear in UI

### ISSUE THREAD

**Issue by danr1979** [swagger-api/swagger-ui#3803]
| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  |
| Which Swagger-UI version?       |
| How did you install Swagger-UI? | 
| Which browser & version?        |
| Which operating system?         | 

### Demonstration API definition

```yaml
responses:
  '200':
    description: OK
    content:
      application/json:
        schema:
          oneOf:
            - $ref: '#/components/schemas/AuthorisationResponse'
            - $ref: '#/components/schemas/ForbiddenResponse'
```

### Configuration (browser query string, constructor, config.yaml)

```js
constructorConfig contains just the default values.
```
### Expected Behavior
Since I have two schemas passed in as a $ref to the 200 response I would expect to see two schemas in the Swagger UI application.  

### Current Behavior
Instead I see only the reference to the 200 response and its description with no example responses displayed as JSOn in the black boxes.

**Comment by webron** [comment]
You've left out a bunch of details from the table - please fill them in.

We also need a full API definition - we can't really test the issue with the snippet you provided.

**Comment by mrichar1** [comment]
The following spec shows nothing in the responses section of the UI - replacing the oneOf section with just an object containing either of the refs and it works fine.

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  | 3.0.0
| Which Swagger-UI version?       | 3.4.2
| How did you install Swagger-UI? | CDN (Cloudflare)
| Which browser & version?        |  firefox 56/Chromium 62
| Which operating system?         | Linux (Ubuntu 16.04.3

``` json
{
  "openapi": "3.0.0",
  "info": {
    "contact": {
      "email": "none@example.com",
      "name": "No Name"
    },
    "description": "Example API",
    "version": "0.0.2",
    "title": "example"
  },
  "paths": {
    "/test": {
      "options": {
        "responses": {
          "200": {
            "content": {
              "application/vnd.api+json": {
                "schema": {
                  "oneOf": [
                    {
                      "$ref": "#/definitions/foo"
                    },
                    {
                      "$ref": "#/definitions/bar"
                    }
                  ]
                }
              }
            }
          }
        }
      }
    }
  },
  "definitions": {
    "foo": {
      "type": "string"
    },
    "bar": {
      "type": "int"
    }
  }
}
```

**Comment by Zakini** [comment]
Similar issue occurs when using oneOf for a request body.

Q | A
-- | --
Bug or feature request? | Bug
Which Swagger/OpenAPI version? | 3.0.0
Which Swagger-UI version? | which ever version is on editor.swagger.io
How did you install Swagger-UI? | using editor.swagger.io
Which browser & version? | Firefox 57.0
Which operating system? | macOS 10.12.6

```yaml
openapi: 3.0.0
info:
  title: Test API
  version: 0.0.1
servers:
  - url: 'localhost:3000/api'
    description: Local API URL.
paths:
  /test:
    get:
      requestBody:
        content:
          application/json:
            schema:
              oneOf:
                - $ref: '#/components/schemas/a'
                - $ref: '#/components/schemas/b'
      responses:
        '200':
          description: OK
components:
  schemas:
    a:
      type: integer
    b:
      type: string
```

**Comment by epetitje** [comment]
Hello,

I have the exact same problem and this issue still exists in version 3.5.

**Comment by Oleg-Arkhipov** [comment]
I would be glad to see this implemented too. In the current state using multiple responses is not really usable because documentation consumer has to dig into YAML/JSON to find out all this information.

**Comment by crussell52** [comment]
@shockey I'm confused why this has been tagged as a `Feature` instead of `Bug`.  According to the [Specification](https://swagger.io/specification/#schemaObject), `schema` is listed as supporting `anyOf`, `oneOf`, and `allOf`.

**Comment by crussell52** [comment]
Applying the `oneOf` to type within `components/schema` **does not** work around this problem.

```yaml
components:
  responses:
    ...
    ServerWriteError:
      description: Write Failed
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/ServerWriteErrorResponse'

  schemas:
    ...
    ServerWriteErrorResponse:
      oneOf:
        - $ref: '#/components/schemas/PartialWriteError'
        - $ref: '#/components/schemas/BaseError'
```

You can **sort of** work around this by declaring `type: object` as a sibling to `oneOf`:

```yaml
components:
  responses:
    ...
    ServerWriteError:
      description: Write Failed
      content:
        application/json:
          schema:
            oneOf:
            $ref: '#/components/schemas/ServerWriteErrorResponse'

  schemas:
    ...
    ServerWriteErrorResponse:
      type: object     # Workaround-ish
      oneOf:
        - $ref: '#/components/schemas/PartialWriteError'
        - $ref: '#/components/schemas/BaseError'
```

 **OR** In the referenced schema component
```yaml
components:
 responses:
    ...
    ServerWriteError:
      description: Write Failed
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/ServerWriteErrorResponse'

  schemas:
    ...
    ServerWriteErrorResponse:
      type: object
      oneOf:
        - $ref: '#/components/schemas/PartialWriteError'
        - $ref: '#/components/schemas/BaseError'
```

The work-around isn't great because it puts an empty object in the response example (which is shown by default), but at least consumers can see the relevant models:
![image](https://user-images.githubusercontent.com/375294/34729219-16c4ed88-f52a-11e7-826e-2a37b69d8592.png)

![image](https://user-images.githubusercontent.com/375294/34729305-51685ed4-f52a-11e7-986b-24c812a82cde.png)

**Comment by shockey** [comment]
Hey @crussell52, I tagged it as a feature because Swagger-UI hasn't/doesn't support `oneOf` responses, as opposed to there being an issue with how a `oneOf` responses implementation renders things.

**Comment by crussell52** [comment]
@shockey That makes sense... thanks for the clarification :)

**Comment by vickysg0210** [comment]
I am facing the same issue. Using oneOf in requestBody is fine. You can not see the exact example but can still see the model. User can dig inside. 

However, responses not working. :) Just to share

**Comment by JinSung-Hwang** [comment]
i`m facing the same issue

**Comment by jasongonzales23** [comment]
+1

**Comment by PavelStefanov** [comment]
@shockey I want to fix this issue. But I don't know how to do. There are two cases:
1.
``` yaml
schema: 
  type: array 
  items:  
    oneOf:  
      - type: string  
      - type: integer
```
I can do like this

![image](https://user-images.githubusercontent.com/13963486/35776445-9c6c6dfc-09ad-11e8-92c9-bb391b5a2983.png)

2.
``` yaml
schema:
  oneOf:
  - $ref: '#/components/schemas/a'
  - $ref: '#/components/schemas/b'

components:
  schemas:
    a:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
    b:
      type: array
      items:
        type: integer
        format: int64
```
I can do like this
![image](https://user-images.githubusercontent.com/13963486/35776447-ab853850-09ad-11e8-8237-4714f35f7e38.png)

Or something more complicated like in model tab? What do you think?

**Comment by JinSung-Hwang** [comment]
i want 2 type

**Comment by DustinOgan** [comment]
of those I prefer the second one.  Since we can name examples in the yaml, would it be a stretch to  have
example1:
text box

example2:
text box

**Comment by PavelStefanov** [comment]
@DustinOgan this is just two different case for oneOf attribute, for array and for object, and example like I can do.

**Comment by NeoDobby** [comment]
For arrays I like the approach to have one example of each item type in the example array.
For objects, I think an approach like in the model tab seems appropriate to me.

**Comment by mikefidel** [comment]
I too wish to see some way for Models to display oneOf.  It is used in my OpenAPI document.  It's not used in a particularly critical API endpoint, but thumbs-up to anybody working on a solution.

**Comment by rickyblaha** [comment]
@mikefidel Models in Swagger UI does display oneOf now. We just recently migrated our docs from swagger 2 to open api 3 just for that reason. See this help page: https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/

**Comment by mrichar1** [comment]
@rickyblaha The spec supports it, but not the UI (unless you're saying bug has been resolved in the UI code?)

**Comment by rickyblaha** [comment]
@mrichar1 this bug is not resolved (using oneOf for multiple responses), but reading Mike's post it looked like he was speaking of Models in general, which does work in the latest UI (e.g. a single response schema that refers to a component schema which uses oneOf or anyOf to refer to other component schemas.)

**Comment by mikefidel** [comment]
For clarification: I was referring to Swagger-UI displaying the onOf alternate schemas in the Responses area as either an Example Value or Model or both.  

Here is what Swagger-UI 3.11.0 displays and its associated OpenAPI 3.0 specification fragment.  You'll see nothing displayed.  

![swagger-ui 3 11 0](https://user-images.githubusercontent.com/20869677/36751628-8c6f517a-1bb5-11e8-9ec4-41e0f1a21bb1.png)

![openapi 3 0 spec](https://user-images.githubusercontent.com/20869677/36751635-92bc95ce-1bb5-11e8-9d30-1cf26a20fa29.png)

Again, this is not a hot issue for me -- just something that I wanted to go on record.  Perhaps I added the comment into the wrong issue.

**Comment by webron** [comment]
What would you hope to see there?

**Comment by mgraciano** [comment]
If I can add my 2c here, I think it would be nice to have a select component with all the possible models, and as each one is selected it can show the example and model in the same way. It gives the power to map multiple responses schemas and the user can inspect each one a time.

**Comment by mikefidel** [comment]
The internet has been down all day due to a storm, and frankly I need a couple days to think through the problem.  The eventual answer needs to cover oneOf and anyOf.  It's possible that there exists schema that has nested oneOf and/or anyOf's -- that certainly makes the processing and UI exquisitely nasty.  I need another day or two to think through and answer your simple question, @webron.  I agree with @mgranciano's response and I suspect it is adequate for the lion's share of cases.  However, I have not thought through the problem well enough to believe it works well when oneOf and/or anyOf's are nested at all depths of recursion.  I'm new to React and Redux, but from what little I've learned it seems to be well suited to dealing with this sort of problem.  Will get back to you in a couple days.

**Comment by mikefidel** [comment]
After due consideration, I could not arrive at an acceptable UI.  Of course, if the one assumes the api specification is: 1) simple 2) has relatively few such named attributes, and 3) doesn't nest oneOf's and/or anyOf's I did manage to come up with "on the back of the napkin" UX design that I think would be acceptable.  It fell into the scenario @mgranciano describes or my simple oneOf schema.  However my draft UX falls apart when oneOf itself includes an attribute containing another oneOf or anyOf (or visa-versa).

Just as a comment: I wonder if we could approach the solution differently.  Perhaps the Example Value | Model can highlight that oneOf and/or anyOf applies by outputting a different styles color or background-color.  Being alerted, the user can work through the implications.  Just offering that as an idea.

**Comment by Fatflo** [comment]
+1 this is something I would need in the UI

**Comment by webron** [comment]
For those of you who didn't notice, @PavelStefanov has submitted #3803 as a potential solution for this feature.

Everyone's input on this one has been great and much appreciated. Looks like @mikefidel managed to cover the major concerns I've been having all along. Let's try to summarize this and see how to proceed.

For clarity - this issue is relevant for OAS3 (and later) only.

The issue of displaying examples comes in several places:
- Parameters (since they now support `schema`)
- RequestBody (as a replacement to `body` parameters)
- Responses
- Headers (just like parameters)

While the two obvious places are responses and request body, we cannot ignore the other two places. The solution needs to fit all, otherwise we risk pushing the problem down the line.

The parameters and request bodies representation should also fit into the try-it-out functionality, meaning it should serve as an easy way to provide the input.

The issue also comes in two forms:
- Generating examples when none are provided by the user (which is this ticket).
- Displaying multiple examples when they are provided by the user (This is trying to be covered by #4092, which is based on #3616).

These two are related, because we would display one or the other, but not both. The experience should be similar as they serve the same purpose - showing an API consumer what they're expected to send/receive to/from the API.

`anyOf` and `oneOf` (and `allOf` for that matter) - aren't very meaningful when it comes to *documentation* (not functionality). We shouldn't assume that a consumer of the documentation automatically understands JSON Schema just because it's what the specification supports. The example in #4368 shows this problem exactly - it contains a `oneOf` of 4 items, but if you look at the generated examples, you'd notice that providing the last of the 4 as a value - it would actually **fail**, because it conflicts with the first one. This kind of distinction would be difficult on consumers to understand simply by seeing `oneOf` at the top of the examples. (for those of you who missed it - the last option would also validate against the first schema - this means it does not fit just *one* of the schemas and so it would fail validation. This wouldn't be the case with `anyOf`.) Similarly - `allOf` does not mean merge.

@mikefidel brought up the biggest issue that concerns me - having anyOf/oneOf deeper in definitions, and not just directly under `scheme` - that's where things get really nasty, and at the moment I'm not sure there's a clean solution to that.

It does seem that at the moment the majority of users is interested in supporting oneOf/anyOf at the root of the schema simply to be able to describe different basic input/output options. We could say that for the immediate future we'll try to add support for those cases and punt on supporting the deeper nested cases - I'm ok with that.

For input - I tend to favor either a tabbed view or a drop-down, where the user can choose which one they want to be displayed and served as a baseline for input. For output - we can go with either that or a flat view as proposed in the PR. The down side of a flat view is that it can make the UI very 'long'. On the other hand, it's easier to view all the information. The downside of a tabbed view is dealing with many examples that don't necessarily fit the view, and then you have another scroll-like feature to move between the different tabs.

It's a bit of a long summary, but I hope this has given you some insight into the challenges in supporting this feature. I'm sure any individual would love to see _something_ that supports their use case, but it is our challenge, as project maintainers, to address the community as a whole. We'd appreciate your input and feedback on this based on the summary. I will try to get a UX person involved in this as well, but any additional input you may have will help.

**Comment by mikefidel** [comment]
Excellent critique!

Tabbed or drop-down view:  I have no particular preference regarding tabbed or drop-down views.  My case may be unique, however the alternative data structures are relatively small -- perhaps 10-15 lines of code one way versus the other.

I thought more about the deeply nested alternative data structures.  Perhaps the depth can be controlled through a configuration setting (oh yes... another one of those!).  Make its default setting a reasonable recursion depth, yet allows the user to increase it if they have one of those nightmare schema we talked about.  Decades ago, I wrote a programming language parser that had to cater for copy files within copy files within copy files -- and yet more copy files.  It's amazing what people will code if given half the chance...

Keep up the good work!

**Comment by Dantemss** [comment]
There is at least one case where nested `anyOf`/`oneOf`/`allOf` would be necessary or at least make it a lot easier to use swagger-ui without redesigning the schemas: anything based on JSON API. The JSON API JSON schema is at http://jsonapi.org/schema and includes plenty of these.

**Comment by kritika0205** [comment]
And update on this as of now? I still cannot see my oneOf in the Swagger UI example/model,.

**Comment by cmavromoustakos** [comment]
Same here. This is a blocker for us as we are sending these docs for our api to a customer and they can not display them properly.

**Comment by Nur777** [comment]
+1 
Today I also approached the need for this function. :)

**Comment by PavelStefanov** [comment]
Guys I want to suggest you try to use examples as workaround
```yaml
content: 
  application/json: 
    schema: 
      anyOf: 
        - $ref: '#/components/schemas/Jessica' 
        - $ref: '#/components/schemas/Ron'
 examples: 
  Jessica: 
    value: 
      id: 10 
      name: Jessica Smith 
  Ron: 
    value: 
      id: 20 
      name: Ron Stewart 
```

**Comment by webron** [comment]
Everyone - asking whether it's already implemented or adding +1 comments is not helpful. If we haven't updated that it's implemented, then it's not been implemented, no need to ask. +1'ing just adds more notifications - you can use the reaction feature on the top comment to +1 and show us that this is important to you. 

Nobody saying this is not important - however, there are real issues implementing support for it and until we resolve those there's nothing to implement. If you have any thoughts on how to address the concerns (and they are all written above), feel free to join the conversation. However, don't expect us to make a solution that works _for you_ if it's "good enough", but doesn't address the issues as a whole. We're not going to replace one issue with a set of others.

**Comment by Hrafnkellos** [comment]
Here is an example of how i'm using examples to display multiple error responses from my service
Thank you for your example @PavelStefanov ^^

``` javascript
 responses:
      '400':
        description: "Invalid input"
        schema:
          $ref:  '#/definitions/HttpError'
        examples:
          ServiceClosed:
            message: "Today is not a working, service closed"
            error: "ERROR_SERVICE_CLOSED"
            errorCode: 2
          ServiceClosedEvening:
            message: "service closed in the evening"
            error: "ERROR_SERVICE_CLOSED_EVENING"
            errorCode: 3


definitions
  HttpError:
    required:
      - message
    type: object
    properties:
      message:
        type: string
      error:
        type: string
      errorCode:
        format: int64
        type: integer

```

**Comment by penja** [comment]
+1

**Comment by xeroc** [comment]
+1

**Comment by AlexNik** [comment]
+1

**Comment by Vetrinus** [comment]
+1

**Comment by allnulled** [comment]
Hello, people. The same is happening to me today. I just came to say hello, you know.

**Comment by anxolerd** [comment]
Hi. Encountered this issue today.

Is there anything I can help with to get the issue fixed earlier? 
/cc @PavelStefanov ?

**Comment by Tamriel** [comment]
Another "workaround" is to use [ReDoc](https://github.com/Rebilly/ReDoc) instead of swagger-ui.

**Comment by pascalazenteam** [comment]
> @shockey I want to fix this issue. But I don't know how to do. There are two cases:
> 1.
> 
> ```yaml
> schema: 
>   type: array 
>   items:  
>     oneOf:  
>       - type: string  
>       - type: integer
> ```
> 
> I can do like this
> 
> ![image](https://user-images.githubusercontent.com/13963486/35776445-9c6c6dfc-09ad-11e8-92c9-bb391b5a2983.png)
> 
> 
> ```yaml
> schema:
>   oneOf:
>   - $ref: '#/components/schemas/a'
>   - $ref: '#/components/schemas/b'
> 
> components:
>   schemas:
>     a:
>       type: object
>       properties:
>         id:
>           type: integer
>           format: int64
>         name:
>           type: string
>     b:
>       type: array
>       items:
>         type: integer
>         format: int64
> ```
> 
> I can do like this
> ![image](https://user-images.githubusercontent.com/13963486/35776447-ab853850-09ad-11e8-8237-4714f35f7e38.png)
> 
> Or something more complicated like in model tab? What do you think?

Hello,
Not sure this is still valid... but as we met the same issue (2018-12)

In our case it would be sufficient to use the second option in your proposal. Adding a line before and after the ','

A clear display would be to replace the ',' by a logical operator (XOR, OR , AND) but this may require to get separate 'example'  display - one for each $ref of the oneOf (XOR) or anyOf (OR) or  allOf (AND)

**Comment by zyv** [comment]
@webron is there anything we can do to help you to get this issue fixed? Is there a BountySource account for swagger-ui / anyone willing to implement this? We would be happy to chip in...

**Comment by penja** [comment]
➕

**Comment by verschoof** [comment]
I'm not sure what the struggle is.

@PavelStefanov I think you should repeat type 1 so you see its a new response, type 2 is confusing since is looks like it could be one response.

Sadly i'm not able to make a PR for this. If somebody can make this? Something is better then nothing. (the milestone is from 8 December 2017 😞 )

**Comment by chrysthynee** [comment]
I'm facing the same issue, can anybody tell me if they have found a workaround to see the examples as well?

**Comment by wol190** [comment]
> Another "workaround" is to use [ReDoc](https://github.com/Rebilly/ReDoc) instead of swagger-ui.

Thanks for the pointer. It's interesting to see another presentation alternative. Example responses using ReDoc still have problems though. For example ([OneOfExample1.yaml.txt](https://github.com/swagger-api/swagger-ui/files/2953209/OneOfExample1.yaml.txt)), with the following schema in the response:
```
OneOfExample1:
  type: object
  oneOf:
    - required:
      - a
    - required:
      - b
  properties:
    a:
      type: integer
    b:
      type: integer
```
ReDoc offers an invalid json example (should be either "a" or "b", but not both):
```
{
  "a": 0,
  "b": 0
}
```
However, the Swagger UI is less helpful

![oneOf](https://user-images.githubusercontent.com/8318503/54140723-2d391700-441c-11e9-8ce9-1dc204e02ed2.jpg)

A workaround is to define the full schema in each case
```
OneOfExample1:
  type: object
  oneOf:
    - required:
        - a
      properties:
        a:
          type: integer
    - required:
        - b
      properties:
        b:
          type: integer
```
![oneOf2](https://user-images.githubusercontent.com/8318503/54141291-48f0ed00-441d-11e9-9612-484760abb117.jpg)

However, I'd prefer to be able define the properties of "a" and "b" outside of the oneOf and still be given more indication of valid example(s) for the schema. The resolution of this issue will be interesting.

**Comment by jemerald** [comment]
@wol190 your sample model does not suggest exclusiveness of `a` and `b`, it just suggests that one of the two field is required. So the example rendered is actually correct.

Just checked with the latest ReDoc, and it also renders the contract correctly:

![image](https://user-images.githubusercontent.com/9168432/55763897-281fc400-5abd-11e9-9bf9-11aa48e163cb.png)

![image](https://user-images.githubusercontent.com/9168432/55763907-3241c280-5abd-11e9-8f0e-957d5c95df11.png)

**Comment by GregoryPellegrin** [comment]
It works fine if you do this
However there is no Example Value, only Schema

```
openapi: 3.0.0
info:
  title: Test API
  version: 0.0.1
servers:
  - url: 'localhost:3000/api'
    description: Local API URL.
paths:
  /test:
    get:
      requestBody:
        content:
          application/json:
            schema:
              type: object
                oneOf:
                  - $ref: '#/components/schemas/a'
                  - $ref: '#/components/schemas/b'
      responses:
        '200':
          description: OK
components:
  schemas:
    a:
      type: integer
    b:
      type: string
```

**Comment by wol190** [comment]
> @wol190 your first schema does NOT suggest exclusiveness of `a` and `b`, it just suggests that one of the two field is required. So the example rendered is actually correct.

@jemerald Many thanks for your response. My conclusion was based on pasting the schema (below) into https://www.jsonschemavalidator.net/ (e.g. Selected Schema: Schema Draft v7):
```
{ 
  "type": "object",
  "oneOf": [
    {
      "required": [ "a" ]
    },
    {
      "required": [ "b" ]
    }
  ],
  "properties": {
    "a": {
      "type": "integer"
    },
    "b": {
      "type": "integer"
    }
  }
}
```
Then providing the following json to test against the schema (Input JSON in https://www.jsonschemavalidator.net/):
```
{
  "a": 0,
  "b": 1
}

```
gives the following error: JSON is valid against more than one schema from 'oneOf'. Valid schema indexes: 0, 1. This was my basis for assuming exclusiveness of `a` and `b`.

**Comment by wol190** [comment]
@jemerald The use of examples with respect to ReDoc is an excellent suggestion - I really like the resulting rendering. Thank you for the pointer. 

With regards to the example, that's based on examples provided within the 3GPP mobile telephony specifications (e.g. clause 5.3.14 of 3GPP TS 29.501) - hopefully therefore providing some motivation for tooling.

**Comment by webron** [comment]
<Similar update to https://github.com/swagger-api/swagger-ui/issues/2651#issuecomment-488335500 for those of you who follow both tickets>

Everyone, some good news on this ticket (at last!). We've been working on a solution for oneOf rendering over the last couple of weeks, and are in the final stages of reviewing the changes, making sure they will work well with the try it out functionality and provide all the needed functionality. 

There's a tiny gotcha - we're also working on a facelift for the existing UI, taking in a lot of the feedback we've received over the years and giving you a nicer looking, more functional product. The design of this solution is based on that facelift (more information on it will be available in the README in the upcoming weeks).

The good news is that we're not going to wait for the facelift work to be done before implementing it, but rather to the best of our ability back port the functionality into the current design.

We hope to share mocks with you as soon as next week (but might be delayed another week) and look forward to seeing your feedback. When we share the mocks, I'll do my best to explain how support for `anyOf` and `not` will also be added (on top of `oneOf`), and why solution in ReDoc (which is a very nice solution) doesn't fit our use case. The implementation itself will come _after_ the [multiple examples](#2651) is done, but is prioritized high right after it.

We understand how frustrating it's been waiting for the implementation, and appreciate the patience, feedback and support from everyone. This is a matter of weeks now (and not referring to hundreds or dozens of those, much much less :wink:).

**Comment by mikefidel** [comment]
That is good news indeed!

**Comment by GregoryPellegrin** [comment]
Can't wait tout see this 🤗🤗

**Comment by ppazos** [comment]
Instead of just a rendering problem, I'm getting an error on the format:

```
responses:
        '200':
          description: Query result strutures ....
          schema:
            oneOf:
              - $ref: '#/definitions/Query_Result_Datavalue'
              - $ref: '#/definitions/Query_Result_Composition'
```
Error:
schema should not have additional properties
additionalProperty: oneOf

That seems related to this issue.

**Comment by jemerald** [comment]
@ppazos 

judging from the fact you are referencing `#/definitions/` instead of `#/components/schemas/`, you are using Swagger 2.0 instead of OpenAPI 3.0.

`oneOf` is only available with OpenAPI 3.0, so you'll need to migrate your schema to the new version.

**Comment by ppazos** [comment]
@jemerald thanks, is there a guide for the migration? I'm using the online editor.

**Comment by ppazos** [comment]
@jemerald sorry, I'm using the hub, not the editor. In the editor found an edit function to do the conversion. That function doesn't seem available on the Hub editor, at least I can't find it.

**Comment by webron** [comment]
@ppazos - https://app.swaggerhub.com/help/openapi-3-support look under the header `Convert OpenAPI 2.0 APIs to OpenAPI 3.0`.

**Comment by h4t0n** [comment]
Waiting for this

**Comment by Alam11** [comment]
I am still facing the issue. Is this implemented in swagger editor ? Thanks for the hard work :)

**Comment by Moro-Code** [comment]
Is there any clear date on when this will be implemented?

**Comment by shockey** [comment]
@Moro-Code no promises on an ETA, but this is currently the next major item for the core team to address 😄

**Comment by Moro-Code** [comment]
Nice! Only Pestering because this would make a world of a difference for me and I'm sure many others. Thanks for the hard work!

**Comment by BryceDoganer** [comment]
I'm also anxiously waiting for `oneOf` to be rendered in the UI.  :P

**Comment by RamiaSaidawi** [comment]
thanks for your work, hope it is done soon :)

**Comment by shockey** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[TEST] Multiple responses using oneOf attribute do not appear in UI](https://trello.com/c/LQ4lOdCx/1-test-multiple-responses-using-oneof-attribute-do-not-appear-in-ui)

**Comment by shockey** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[TEST] Multiple responses using oneOf attribute do not appear in UI](https://trello.com/c/LQ4lOdCx/1-test-multiple-responses-using-oneof-attribute-do-not-appear-in-ui)

**Comment by jtrees** [comment]
Any updates?

**Comment by shockey** [comment]
Folks: per the [Swagger Contribution Guide](https://github.com/swagger-api/.github/blob/master/CONTRIBUTING.md#voting), please refrain from making comments that only serve to bump the issue or indicate interest — use Reactions to vote for issues instead, since we use them [to sort issues based on user interest](https://help.github.com/en/articles/sorting-search-results#sort-by-reactions).

Thanks!

**Comment by amandajordaan** [comment]
I have this exact problem. Has anyone found a work-around for this? I need both my examples to show. Currently my models show beautifully, but my example values are blank. 

`                "schema": {
                  "oneOf": [
                    "$ref": "./components/schemas/responses/schedules/postgroupschedule.json",
                    "$ref": "./components/schemas/responses/schedules/postindividualschedule.json"
                    ]
                }
              }
`

**Comment by iget-master** [comment]
No ETA for this yet? More than 2 years after first issue opened...

**Comment by eLvErDe** [comment]
If you look at the commits you will see SwaggerUI development is pretty much dead. I don't think we can expect this feature until a take over occurs...

**Comment by webron** [comment]
@eLvErDe It's not dead, but unfortunately right now we're understaffed and doing our best to address more immediate issues. Hopefully things will pick up soon and we'll be able to provide the community with even more.

**Comment by iget-master** [comment]
@webron give power to community at least to approve pull requests...

**Comment by webron** [comment]
You should be able to approve/review PRs (not merge them). If that's not the case, please let me know and I'll look into it.

**Comment by eLvErDe** [comment]
@webron well saldy this is GitHub, not a company meeting. I don't think anyone is buying marketing talks here. The fact is this ticket has been around for nearly 3 years and the conclusion is that SwaggerUI is still not handling OpenAPI 3 correctly. So yes, it's a dead project, period'

**Comment by eLvErDe** [comment]
@iget-master do you have a PR implementing anyof correctly ? 😍

**Comment by eLvErDe** [comment]
Well yeah there's one at least:
https://github.com/swagger-api/swagger-ui/pull/5530

Time to build my own version :)

**Comment by iget-master** [comment]
Yes, there are 67 open pull requests. At least 10 are ready to merge.

**Comment by eLvErDe** [comment]
In case something want a build in the meanwhile:
https://github.com/eLvErDe/swagger-ui/releases

It's working perfectly fine with my Python module, dist folder include built js/css, based on 3.24.3 tag with only PR#5530 included.

**Comment by iamjoshuabcxvii** [comment]
Almost 3 yrs later. This is still an issue for me in the SwaggerUI. any update about a fix in the UI rendering?.

**Comment by iamjoshuabcxvii** [comment]
Almost 3 yrs later. This is still an issue for me in the SwaggerUI. any update about a fix in the UI rendering?.

**Comment by NilsSchn** [comment]
Hi,

have the same problem here.
Downloaded the current master branch of swagger ui, but the example rendering, while using oneOf,  still does not work.

Would be nice, if there will be a fix soon.
Maybe someone knows a workaround or something else?

Thanks, Nils.

**Comment by silminawaz** [comment]
encountered same issue, and our APIs use a 200 status code with an inner status that indicates error and success.. so I need to display to the user the vsrious use cases and cannot do this. It would be great if it gets fixed.

**Comment by Gibbo3771** [comment]
Still a problem, new to Swagger so it could be a YAML structure problem. However, if I have a `requestBody` that can contain a key of `params` and those params could be `oneOf` many different schemas, the entire thing is useless since the examples do not show.

**Comment by TakeawayInnovation** [comment]
Seriously wonder why the pull request still has no official response from the maintainers of this project even though it's been there for months. Please look at it 😢 . A lot of people are waiting on this.

**Comment by Ryangr0** [comment]
Seriously wonder why the pull request still has no official response from the maintainers of this project even though it's been there for months. Please look at it 😢 . A lot of people are waiting on this.

**Comment by umutmaz** [comment]
Having the same issue in  requestBody. Using oneOf to point multiple schemas and in swaggerUI; examples do not show.

**Comment by hbensalem** [comment]
I'm having the sane issue with anyOf in request body

**Comment by mathis-m** [comment]
Swagger ui should generate multiple examples for nested combined schemas as well.

For example:
```yml
openapi: 3.0.3
info:
  title: title
  version: 0.0.0
paths:
  /test:
    post:
      summary: test
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/nested'
      responses:
        200:
          description: OK
      operationId: test.post
components:
  schemas:
    nested:
      type: object
      properties:
        some: 
          type: string
        nestedProp:
          type: object
          anyOf:
            - $ref: '#/components/schemas/one'
            - $ref: '#/components/schemas/two'
    one:
      title: One
      type: object
      required:
        - name
      properties:
        name:
          type: string
          pattern: ^one$
      example: 
        name: one
    two:
      title: Two
      type: object
      required:
        - name
      properties:
        name:
          type: string
          pattern: ^two$
      example:
        name: two
```
is rendered like:
![image](https://user-images.githubusercontent.com/11584315/92290445-61f43480-ef14-11ea-8705-7e259ad1e490.png)
and should be rendered like:
```yml
openapi: 3.0.3
info:
  title: title
  version: 0.0.0
paths:
  /test:
    post:
      summary: test
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/nested'
            examples:
              nested_One:
                $ref: '#/components/examples/nested_One'
              nested_Two:
                $ref: '#/components/examples/nested_Two'
      responses:
        200:
          description: OK
      operationId: test.post
components:
  examples:
    nested_One:
      value: 
        some: string
        nestedProp:
          name: one
    nested_Two:
      value:
        some: string
        nestedProp:
          name: two
  schemas:
    nested:
      type: object
      properties:
        some: 
          type: string
        nestedProp:
          type: object
          anyOf:
            - $ref: '#/components/schemas/one'
            - $ref: '#/components/schemas/two'
    one:
      title: One
      type: object
      required:
        - name
      properties:
        name:
          type: string
          pattern: ^one$
      example: 
        name: one
    two:
      title: Two
      type: object
      required:
        - name
      properties:
        name:
          type: string
          pattern: ^two$
      example:
        name: two
```
![image](https://user-images.githubusercontent.com/11584315/92290744-7be24700-ef15-11ea-88d7-9012933066b6.png)

**Comment by ssmshah1** [comment]
+1

**Comment by EdgarSedov** [comment]
Well idk how u guys sort issues by emojis but simple check by comments shows this is kinda #1 concrete issue with huge difference from others. And it's been here for 3 years now.
![image](https://user-images.githubusercontent.com/16332716/94935455-c388ce00-04d5-11eb-9f5d-fc670c397c62.png)

Personally i fully understand that it's free time work, no payment. But if u can't keep maintaining this thing, maybe it's time to make an announcement to let someone take over? My point is:  i see fresh docker tags for swagger-ui, i pull them and starting to use. When i see the problems are still here. If only i knew project is dead i wouldn't wasted any time and used redoc. So it's better for saving others people time mark project as dead or let some to take over.

**Comment by jokaorgua** [comment]
any news from maintainers regarding this issue?

**Comment by NagoLazaro** [comment]
Are there any updates from Swagger team regarding this functionality?

**Comment by hpl002** [comment]
+1

**Comment by by-Exist** [comment]
+1

**Comment by raffig** [comment]
Any news about this issue?

**Comment by lsc-dotsource** [comment]
considering the website of SwaggerUI claims "Complete OAS Support Complete OAS Support Visualize APIs defined in Swagger 2.0 or OAS 3.0" and oneOf, anyOf and allOf are clearly part of openApi 3, this should thus be re-categorized as a Bug, or the softwares description be corrected.

**Comment by anthony-janocko** [comment]
Any update here?  Not working for subschemas in responses.

**Comment by Laurianti** [comment]
Any news?

**Comment by jacobjmarks** [comment]
I am having a similar issue here. See below an example spec in which I am attempting to describe a form which can accept one of two possible schemas:

``` yaml
openapi: 3.0.1
info:
  title: FooBar
  version: 1.0.0
paths:
  /foobar:
    post:
      summary: Summary
      requestBody:
        content:
          application/x-www-form-urlencoded:
            schema:
              oneOf:
              - type: object
                properties:
                  Foo:
                    type: string
              - type: object
                properties:
                  Bar:
                    type: string
      responses:
        default:
          description: Description
```

> Please note that the same issue occurs when using a `requestBody/content` type of `multipart/form-data`

This endpoint should expect to receive a body of either `Foo=` or `Bar=`.
This schema causes the following Swagger UI interface to be rendered:

![image](https://user-images.githubusercontent.com/19641427/141227881-e80d3805-af76-4451-9810-3f03fbb75938.png)

This is _not_ correct. It is representing the form as a JSON payload and causes invalid requests to be sent to the endpoint in which the Swagger UI is encoding each character as a separate form value. E.g. sending a value of `hi` would result in the body `0=h&1=i`.

This same issue occurs when using `anyOf`.

When using `allOf`, however, a correct representation is rendered as below:

![image](https://user-images.githubusercontent.com/19641427/141227963-19f881e9-90fb-4bf3-bda8-299bdae3836f.png)

Are there any updates here regarding the support for such `oneOf` and `anyOf` usage? Alternatively; am I incorrectly defining my schema in some way to achieve such usage?

**Comment by jaredbarranco** [comment]
+1 to this - My scenario:
API client should expect return to contain array of objects who's underlying structure changes based on type fields that are optional query params.

**Comment by tomasvts** [comment]
If anyone is having this problem using the latest 4.4.1 version from unpkg worked from me.    
I'm using the standalone version like this:

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Swagger API</title>
    <link rel="stylesheet" type="text/css" href="https://unpkg.com/swagger-ui-dist@4.4.1/swagger-ui.css">
</head>
<body>

<div id="swagger-ui"></div>

<script src="https://unpkg.com/swagger-ui-dist@4.4.1/swagger-ui-standalone-preset.js"></script>
<script src="https://unpkg.com/swagger-ui-dist@4.4.1/swagger-ui-bundle.js"></script>
<script>
    window.onload = function() {
        window.ui = SwaggerUIBundle({
            url: "YOUR URL HERE",
            dom_id: '#swagger-ui',
            deepLinking: true,
            presets: [
                SwaggerUIBundle.presets.apis,
                SwaggerUIStandalonePreset
            ],
            plugins: [
                SwaggerUIBundle.plugins.DownloadUrl
            ],
        })
    }
</script>
</body>
</html>
```

**Comment by EladG77** [comment]
Any news about this?

**Comment by tomasvts** [comment]
Try updating to the last version

**Comment by thiagomengue** [comment]
Hello, 

Does anyone have a solution or workaround ?
I'm using the swaggerapi/swagger-ui:v4.11.0 docker image but the problem persists....

**Comment by ajith428** [comment]
Any update/workaround for the issue?

**Comment by alecnmk** [comment]
👋 the is a workaround I've found working and built a simple demo: 
https://gitlab.com/alecnmk/openapi-swaggerui-multiple-response-examples-demo/-/blob/main/openapi.yaml#/default/get_products__id_

**Comment by WellingtonSI** [comment]
I'd like to thank  [alecnmk](https://github.com/alecnmk) for the help with the code.

**Comment by BetclicDamian** [comment]
After 5 years, I think we can safely assume that this bug will never be fixed

**Comment by vpaflah** [comment]
@jaredbarranco ' solution worked for me.

**Comment by njmulsqb** [comment]
I am also having this issue, @alecnmk solution worked partially, but anyOf is not working at all. 
Using examples is not resolving the example content from YAML file properly.

**Comment by Moro-Code** [comment]
This issue should have been solved https://github.com/swagger-api/swagger-ui/issues/5435#event-8668317571

**Comment by ivanzigoni** [comment]
> 👋 the is a workaround I've found working and built a simple demo: https://gitlab.com/alecnmk/openapi-swaggerui-multiple-response-examples-demo/-/blob/main/openapi.yaml#/default/get_products__id_

looks great but for anyone trying to use nestjs swagger module decorators, this unfortunately doesn't seem to apply

**Comment by fridolinf** [comment]
> 👋 the is a workaround I've found working and built a simple demo: https://gitlab.com/alecnmk/openapi-swaggerui-multiple-response-examples-demo/-/blob/main/openapi.yaml#/default/get_products__id_

I think this could be a reference if someone wants to do a JSON implementation.

### file1.json
```yaml
{
  ...your_rest_implementation,
  "responses": {
    "400": {
      "description": "Client error",
      "content": {
        "application/json": {
          "schema": {
            "$ref": "#/components/schemas/StandartClientError"
          },
          "examples": {
            "FormValidationError": {
              "$ref": "#/components/schemas/StandartClientError/examples/FormValidationError"
            },
            "SimpleClientError": {
              "$ref": "#/components/schemas/StandartClientError/examples/SimpleClientError"
            }
          }
        }
      }
    },
  }
```


### file2.json
```yaml
{
  "StandartClientError": {
    "type": "object",
    "properties": {
      "success": {
        "type": "boolean",
        "example": false
      },
      "message": {
        "type": "string",
        "example": "Client error"
      }
    },
    "examples": {
      "FormValidationError": {
        "value": {
          "success": false,
          "message": {
            "formKey": [
              "Invalid data"
            ]
          }
        }
      },
      "SimpleClientError": {
        "value": {
          "success": false,
          "message": "Client error"
        }
      }
    }
  }
}
```

**Results:**
<img width="577" alt="Image" src="https://github.com/user-attachments/assets/165ffa6b-fbb9-4cfd-9a73-1b6bddec8551" />

<img width="557" alt="Image" src="https://github.com/user-attachments/assets/1a6eafb0-6eb5-4c4a-804b-a8a983e7dd13" />
